Cleaning models

iamjwc opened the pull request:
Hey,

I noticed that some of the model code was actually in the top level Server class, so I decided to fork the project and move that code into a Model parent class. All of the tests are still passing (although, there don't seem to be 'unit' tests).

Also, I noticed that the Server.create_indexes method is being called to add blocks to an array that would create the mongo indices, but then Server.create_indexes never gets called to actually execute those blocks. I separated the two types of .create_indexes calls into .add_indexes and .create_indexes. All the blocks that were .create_indexes are now .add_indexes, so .create_indexes still needs to be called from somewhere.

Justin

iamjwc committed:
Moved modely features out of the Server class

iamjwc committed:
Missed that one

iamjwc committed:
Moved Model class out of Server class

iamjwc committed:
Model is singular, so I renamed the file

iamjwc committed:
Moved finder to MongoHelpers module

iamjwc committed:
Removed tabs

iamjwc committed:
Cleaned up AccessToken methods

iamjwc committed:
Refactored AuthRequest class methods

iamjwc committed:
Refactored Client creation

iamjwc committed:
Cleaned up AccessGrant

iamjwc committed:
Unified create/document interface

iamjwc committed:
Moved .create to Model

iamjwc committed:
Moved allowed_scope helper out of all classes and into the client

iamjwc committed:
Moved common fields into parent Model

iamjwc committed:
Renamed get_token_for to find_or_create

iamjwc committed:
Got rid of special .new method on Model

assaf commented:
When using ROS, there's only one API you need to care about, which is encapsulated in the class `Server`, hence setting `Server.database`.

The method name `get_token_for` tells me what it does, the method name `find_or_create_by_client_and_scope_and_identity` tells me how it does it. I much prefer to deal with the "what it does" abstraction.

Seems this remove access to the attributes `created_at` and `revoked`.

When you load a MongoDB record (`find` or `find_one`) the primary key maps to the hash key `_id`, so the models have an `_id` attribute to match.

There seems to be changes to the data schema, so this will break existing installations.

iamjwc commented:
Hey Assaf,

I moved the `Model.database` that I changed back to `Server.database` to provide a singular API to deal with. I wasn't looking at that when I made the change initially, but that makes more sense than what I had.

I also changed the method names back to `#get_token_for`.

The attributes `created_at` and `revoked` were actually moved up to the parent class because they are common across all models. I wrote tests that ensure that they are accessible and correctly set.

I also added specs around `.create` and `.find` that ensure that `_id` is correctly set. Also, this verifies that the aliases that were made for `_id` are correct.

I really believe there are no schema changes to this refactoring. I am only moving things that are already there around.

Hope this helps. Please let me know if you have other concerns and I can address them.

Thanks,

Justin

assaf commented:
'''
The attributes created_at and revoked were actually moved up to the parent class because they are common across all models. I wrote tests that ensure that they are accessible and correctly set.
'''

See the problem? When I looked into Client it didn't see created_at. I had to know to look into Model, and then mentally combine two lists to come up with the actual attribute list. Originally I was just being lazy: listing all the methods/attributes in each model makes the code easier to maintain.

Another consideration, if someone decides to rewrite the models for SQL or Redis, for example, they can look into a class like Client or AccessToken and immediately know what they have to implement: everything that's listed there. They don't have to re-implement any of the methods in Server/Utils, although they could use them it they wanted to.

I don't see the shared class Model as adding much, but I can certainly see it as taking away from the simplicity of the current design.
